server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/NodeAllocationOrdering.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class NodeAllocationOrdering {
+
+    private final AtomicLong order = new AtomicLong(0);
+    private final Map<String, Long> recentAllocations = new HashMap<>();
```

Comment:
I wonder if we should clear this map from no longer existing nodes more aggressive then just on master change? 
If so do we care about nodes leaving cluster for a short period of time?